ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java

```diff
@@ -709,6 +673,69 @@ private void processReplicaMetadataResponse(Set<StoreKey> missingRemoteStoreKeys
+        // deleted yet locally
+        MessageInfo localMessageInfo = remoteReplicaInfo.getLocalStore().findKey(localKey);
+        boolean deletedLocally = localMessageInfo.isDeleted();
+        boolean ttlUpdatedLocally = localMessageInfo.isTtlUpdated();
+        short localLifeVersion = localMessageInfo.getLifeVersion();
+        short remoteLifeVersion = messageInfo.getLifeVersion();
+        if (localLifeVersion > remoteLifeVersion) {
+          // if the lifeVersion in local store is greater than the remote lifeVersion, then nothing needs to be done.
+          continue;
+        } else if (localLifeVersion == remoteLifeVersion) {
+          // we are operating in the same version, in this case, delete would be the final state.
+          if (!deletedLocally) {
+            // Only adds record when it's not deleted yet. Since delete is the final state for this lifeVersion, if there
+            // is a delete record for the current lifeVersion, then nothing needs to be done.
+            MessageInfo info = new MessageInfo(localKey, 0, localKey.getAccountId(), localKey.getContainerId(),
+                messageInfo.getOperationTimeMs());
+            if (messageInfo.isTtlUpdated() && !ttlUpdatedLocally) {
+              applyTtlUpdate(info, remoteReplicaInfo);
+            }
+            if (messageInfo.isDeleted()) {
+              applyDelete(info, remoteReplicaInfo);
+            }
+          }
+        } else {
+          // if we are here, then the remote lifeVersion is greater than the local lifeVersion.
+          // we need to reconcile the local state with the remote state.
+          //
+          // There are three states we have to reconcile: lifeVersion, ttl_update, is_deleted.
+          // To reconcile lifeVersion and is_deleted, we have to add a Delete or Undelete record, based on what the final state is.
+          // to reconcile ttl_update, if the final state is delete, then, we have to add ttl_update before delete, other, we can add ttl_update after undelete.
+          MessageInfo info = new MessageInfo(localKey, 0, localKey.getAccountId(), localKey.getContainerId(),
+              messageInfo.getOperationTimeMs(), remoteLifeVersion);
+          boolean shouldInsertTtlUpdate = false;
+          if (messageInfo.isTtlUpdated() && !ttlUpdatedLocally) {
+            // make a patch for ttl update
+            // if the remote state is delete, then we can't insert TTL_UPDATE after delete, we have to insert a ttl_update here
+            if (messageInfo.isDeleted()) {
+              // since ttl update can only follow Put or Undelete, make sure it's not locally deleted.
+              // we can reuse the lifeVersion for undelete and ttl update, since the delete would be the final state of
+              // this lifeVersion.
+              if (deletedLocally) {
+                applyUndelete(info, remoteReplicaInfo);
```

Comment:
What if U exists in local store?

Reply:
The latest value locally is a delete, and we can't apply ttl_update after a delete, so here I apply a undelete locally first. It doesn't matter if U exists in local store, as long as the latest value is a delete.